### ISSUE TITLE
python setup.py build fails on OpenBSD

### ISSUE THREAD

**Issue by kwatsen** [python-ldap/python-ldap#273]
If you found a bug in python-ldap, or would request a new feature,
this is the place to let us know.

Please describe the issue and your environment here.

---

Issue description:

Steps to reproduce: 

Operating system: 

Python version: 

python-ldap version:

**Comment by kwatsen** [comment]
Note: a fix is to add these two lines to th `[_ldap]` section in `setup.py`:
```
+library_dirs = /usr/local/lib
+include_dirs = /usr/local/include /usr/local/include/sasl
```
After adding the above two lines, then it compiles cleanly.

I don't know if there is a way to add this to `python-ldap/Build/build-openbsd`, as I think that file is more for compiling in the OpenBSD `ports` tree (not for, e.g., `pip install`).

Happy to submit a PR if the devs want to suggest a good multi-platform fix.

**Comment by tiran** [comment]
I'm not that familiar with OpenBSD. Isn't there a better way to configure the compiler, linker, and dynamic loader to pick up these locations by default?

Python just calls the compiler and linker. You can also set LDFLAGS and CFLAGS env vars.

**Comment by kwatsen** [comment]
Wait, no, that can't be right.  `pip install` should just work.  Yes, I've debugged it some by cloning the git repo and running `python setup.py ...`, but most wouldn't...

I think the needed fix is in the `setup.py` file, around where it says: `#-- Read the [_ldap] section of setup.cfg`.  My guess is that you want something like `if platform.system() == 'OpenBSD': ...`

Ideally, the details could stay in the setup.cfg file:
```
[_ldap_openbsd]
library_dirs = /usr/local/lib
include_dirs = /usr/local/include /usr/local/include/sasl
```

and then the setup.py just brings this section in when platform is "OpenBSD"?

**Comment by tiran** [comment]
I agree that it should just work but I don't agree with your proposed solution. python-ldap is not the correct place to solve a compiler and linker issue. It's either a problem with OpenBSD, compiler settings, OpenBSD's package manager, or Python on OpenBSD. The build system should pick up default locations for headers and libraries.

The misconfiguration affects all Python packages with a C extension, not just python-ldap.

**Comment by gegeweb** [comment]
Same issue with 3.2.0 on FreeBSD (11.2-RELEASE).
No problem to compile the 3.1.0 in a virtualenv on the same system.

**Comment by quanah** [comment]
@tiran I think you're used to building software on Linux.  The BSDs are quite a bit different.  Rather than having thousands of default packages you install in the system path, you generally build them out using a "ports" system.  Ports are installed into "/usr/local".  The default compiler available (clang) does not, on purpose, have /usr/local/include in its include path.  You can see this when executing cc -v on a *BSD system:

```
c -v foo.c
FreeBSD clang version 6.0.1 (tags/RELEASE_601/final 335540) (based on LLVM 6.0.1)
Target: x86_64-unknown-freebsd12.0
Thread model: posix
InstalledDir: /usr/bin
 "/usr/bin/cc" -cc1 -triple x86_64-unknown-freebsd12.0 -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name foo.c -mrelocation-model static -mthread-model posix -mdisable-fp-elim -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -v -resource-dir /usr/lib/clang/6.0.1 -fdebug-compilation-dir /home/build -ferror-limit 19 -fmessage-length 189 -fobjc-runtime=gnustep -fdiagnostics-show-option -fcolor-diagnostics -o /tmp/foo-2f9c89.o -x c foo.c
clang -cc1 version 6.0.1 based upon LLVM 6.0.1 default target x86_64-unknown-freebsd12.0
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/clang/6.0.1/include
 /usr/include
End of search list.
```

So really what this request is, is to ensure that python-ldap can be built and installed for the ports system that is used on BSD Oses (FreeBSD, OpenBSD, possibly NetBSD as well).

**Comment by quanah** [comment]
See also https://www.freebsd.org/ports/

**Comment by encukou** [comment]
I guess the questions is: is hardcoding `library_dirs = /usr/local/lib` and `include_dirs = /usr/local/include /usr/local/include/sasl` really the best option for OpenBSD? Is it documented anywhere? Are there other ways to install OpenLDAP than the ports system?

I woulndn't mind having better, system-specific defaults for settings that aren't overridden in `[_ldap]`.

**Comment by iblis17** [comment]
> I guess the questions is: is hardcoding library_dirs = /usr/local/lib and include_dirs = /usr/local/include /usr/local/include/sasl really the best option for OpenBSD? Is it documented anywhere?

Yes, and Yes. The prefix `/usr/local` is the system default. The source code is here:
- FreeBSD: https://svnweb.freebsd.org/base/head/share/mk/bsd.compiler.mk?revision=354736&view=markup#l59
- OpenBSD (The `LOCALBASE` var in line 11): http://cvsweb.openbsd.org/cgi-bin/cvsweb/ports/infrastructure/mk/pkgpath.mk?rev=1.83&content-type=text/x-cvsweb-markup

 > Are there other ways to install OpenLDAP than the ports system?

Yes, building it manually is for the OpenLDAP hacker/developer, i think. But the standard way is the ports system.

> I woulndn't mind having better, system-specific defaults for settings that aren't overridden in [_ldap].

(Y)

**Comment by sticky-note** [comment]
Resolved with passing environment variable:
```C_INCLUDE_PATH=/usr/local/include```
in the `clang` compiler

**Comment by tiran** [comment]
The [FreeBSD](https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/porting-prefix.html) document discourages hard-coded paths in the upstream sources:

> Avoiding hard-coded /usr/local paths in the source makes the port much more flexible and able to cater to the needs of other sites.

I'm -1 to add ``/usr/local`` to header and library lookup path.

Could somebody with OpenBSD and FreeBSD experience please contribute documentation updates for the installing section instead?
